### PR TITLE
Replace CRA with Vite

### LINKED ISSUE #72: Replace CRA SW with VitePWA SW
**Context**
Currently we use the generated CRA SW. We want to auto-generate it with VitePWA.

We attempted this, but for some reason the `registerSW({immediate: true})` flag does not work. In the `vite.config.ts` we also defined `skipWaiting: true` and `clientClaim: true`. When using the generated SW, when we have two clients locally running the same app, they refuse to update. In the application context it shows that they are still waiting and it requires us to close one of the apps and refresh, or click skipwaiting manually in the console.

This is something to come back to, but as it was taking a lot of time and preventing Vite from being made, it is now a separate issue. 

**Progress branch**
The [AjobK/indico-checkin-pwa/tree/transition-to-vite-pwa](https://github.com/AjobK/indico-checkin-pwa/tree/transition-to-vite-pwa) branch on my own repo is a good baseline to start as I already tried a lot there. I will keep it alive, but keep in mind that it is likely severely outdated (last update on 17/09/2024).

### PR DESCRIPTION
**Pull request by AjobK**
Closes #72 

**Problem**

The compilation process for CRA is long and official support is also being cut off.

**Solution**

Moving to Vite, which is significantly faster!

### CODE REVIEW

File: src/pages/Auth/utils.ts [hunk 1]
```diff
@@ -51,7 +56,11 @@ export function validateEventData(data: any): data is QRCodeEventData {
 
 export interface QRCodeParticipantData {
   serverUrl: string;
-  checkinSecret: string;
+  checkinSecret?: string;
```
Comment (tomasr8) [inline src/pages/Auth/utils.ts:59]: I think that `checkinSecret` should not be optional here

File: src/serviceWorkerRegistration.ts [hunk 2]
```diff
@@ -24,18 +24,10 @@ type Config = {
...
-      // Our service worker won't work if PUBLIC_URL is on a different origin
-      // from what our page is served on. This might happen if a CDN is used to
-      // serve assets; see https://github.com/facebook/create-react-app/issues/2374
-      return;
-    }
-
+  if (import.meta.env.PROD && 'serviceWorker' in navigator) {
     window.addEventListener('load', () => {
-      const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`;
+      const swUrl = './service-worker.js';
```
Comment (tomasr8) [inline src/serviceWorkerRegistration.ts:29]: An absolute path might be a bit safer here

File: vite.config.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import react from '@vitejs/plugin-react';
+import tailwindcss from 'tailwindcss';
+import {defineConfig, loadEnv} from 'vite';
+import viteTsconfigPaths from 'vite-tsconfig-paths';
+import { VitePWA } from 'vite-plugin-pwa';
```
Comment (tomasr8) [inline vite.config.ts:5]: That extra space should have been caught by the linters, but we currently only lint `src/`. We could add this file to the config as well:

```json
{
      "lint": "eslint --ext js,jsx,ts,tsx src/ && prettier --check src/",
}
```

File: src/serviceWorkerRegistration.ts (hunk 2 above)
Comment (AjobK) [reply src/serviceWorkerRegistration.ts]: Agree, fixed and testing by committing more.

File: vite.config.ts (hunk 3 above)
Comment (AjobK) [reply vite.config.ts]: Extended to also check all other `.js`, `.ts` and `.cjs` files that are at the root of the project

**Comment by AjobK** [comment]
**Update:**

Everything works, but the manifest.json is not properly cached to work offline. The rest is cached properly.

**Comment by GovernmentPlates** [comment]
Remember to update the dev instructions in the README! :)

**Comment by AjobK** [comment]
@GovernmentPlates The README is still up-to-date. If you refer to the commands that we use, they are the same as we use the abstraction provided by our `package.json`. The command behind for example `npm start` is changed, but we still use `npm start`.

**Update**
Discussed and confirmed with @GovernmentPlates

**Review by GovernmentPlates — APPROVED** [review]

**Pull request merged**